Enhancements

fmido88 opened the pull request:
- Fix instate of adding new override, editing the old one, so prevent the error of multiple records exists when calling $DB->get_record().
- Adding fields to the database table for time updated and overrider (the person did the override process).
- Adding additional method for overriding by adding additional maxviews instated of complete override.
- Adding capability for overriding, so only users with this capability can perform the override.
- Adding Events for overriding so that manager can track any override process from the log.

I'm looking forward for your reply after reviewing these changes.

fmido88 committed:
Changes to the new version

- Fix instate of adding new override, editing the old one, so prevent the error of multiple records exists when calling $DB->get_record().
- Adding fields to the database table for time updated and overrider (the person did the override process).
- Adding additional method for overriding by adding additional maxviews instated of complete override.
- Adding capability for overriding, so only users with this capability can perform the override.
- Adding Events for overriding so that manager can track any override process from the log.

I'm looking forward for reviewing these changes.

fmido88 commented:
Hello @danielneis
It will be great and means alot for me if you take a look for these changes.

fmido88 committed:
Resolve the mixing between two lang strings

danielneis commented:
Hello, @fmido88 
I'll do my best to review it by the next week.

fmido88 commented:
Thanks @danielneis
I appreciate it very much and waiting to tell me if there is any modifications you need.

danielneis commented on classes/form/override.php:
```diff
@@ -40,33 +44,81 @@ public function definition() {
+
         foreach ($modinfo->cms as $i) {
+            // Filter course modules by that only has maxviews conditions.
+            $getviewslimit = $DB->get_field_select('course_modules', 'availability' ,
```
For sure there is availability information already on $i , it would be nice not to have 1 more query per course module. could please check and change to use the appropriate attribute/function ?

fmido88 replied:
Thanks, I'll use this instate:
```
            if (!empty($i->availability)) {
                $av = json_decode($i->availability);
                foreach ($av->c as $condition) {
                    if ($condition->type !== 'maxviews') {
                        continue;
                    }
                }
            } else {
                continue;
            }
```

danielneis commented on lang/en/availability_maxviews.php:
```diff
@@ -22,24 +22,47 @@
 $string['eithernotdescription'] = 'you have not reached the limit of {$a->viewslimit} views (you have {$a->viewscount} views)';
 $string['eitherdescription'] = 'you have reached the limit of {$a->viewslimit} views (you have {$a->viewscount} views)';
+$string['event_maxviews_override_created'] = 'The limitation of maxviews override created';
+$string['event_maxviews_override_description'] = 'The maxviews of user of id {$a->relateduserid} in course module of id {$a->cmid}
```
please, change the string to read "user with id" , "course with id"

fmido88 replied:
done

danielneis commented on lang/en/availability_maxviews.php:
```diff
@@ -22,24 +22,47 @@
+$string['event_maxviews_override_created'] = 'The limitation of maxviews override created';
+$string['event_maxviews_override_description'] = 'The maxviews of user of id {$a->relateduserid} in course module of id {$a->cmid}
+override {$a->type} by user of id {$a->userid} and {$a->reset}';
+$string['event_maxviews_override_updated'] = 'The limitation of maxviews override updated';
```
please, remove the initial "The"

fmido88 replied:
done

danielneis commented on lang/en/availability_maxviews.php:
```diff
@@ -22,24 +22,47 @@
+$string['event_maxviews_override_description'] = 'The maxviews of user of id {$a->relateduserid} in course module of id {$a->cmid}
+override {$a->type} by user of id {$a->userid} and {$a->reset}';
+$string['event_maxviews_override_updated'] = 'The limitation of maxviews override updated';
+$string['event_maxviews_override_deleted'] = 'The limitation of maxviews override deleted';
```
please, remove the initial "The"

fmido88 replied:
done

danielneis commented on lang/en/availability_maxviews.php:
```diff
@@ -22,24 +22,47 @@
+$string['participant_help'] = 'Select the participants needed to override their maximum views for the selected course module.';
 $string['pluginname'] = 'Max Views';
 $string['resetviews'] = 'Reset views';
+$string['resetviews_help'] = 'If this options checked, the views count for the selected users will reseted to zero for the selected course module.';
```
it should be "will be reset to zero"

fmido88 replied:
done

danielneis commented on lang/en/availability_maxviews.php:
```diff
@@ -22,24 +22,47 @@
 $string['pluginname'] = 'Max Views';
 $string['resetviews'] = 'Reset views';
+$string['resetviews_help'] = 'If this options checked, the views count for the selected users will reseted to zero for the selected course module.';
+$string['resetyes'] = 'maxviews has been rest';
```
we have a typo here, it should be "has been reset"

fmido88 replied:
done

danielneis commented on lang/en/availability_maxviews.php:
```diff
@@ -22,24 +22,47 @@
 $string['resetviews'] = 'Reset views';
+$string['resetviews_help'] = 'If this options checked, the views count for the selected users will reseted to zero for the selected course module.';
+$string['resetyes'] = 'maxviews has been rest';
+$string['resetno'] = 'maxviews not reset';
```
it should be "hasn't been reset"

fmido88 replied:
Done

danielneis commented on classes/form/override.php:
```diff
@@ -40,33 +44,81 @@ public function definition() {
+            $mform->addElement('text', 'maxviews', get_string('maxviews_add', 'availability_maxviews'));
+            $mform->setType('maxviews', PARAM_INT);
+            $mform->addHelpButton('maxviews', 'maxviews_add', 'availability_maxviews');
+        }
```
this could be simpler if you move the calling of addElement, setType and addHelpButton outside of the if condition, and inside you create variables that will have the values used in the function calls after the if/else.
like
 $label = get_string('max_views' ...
$helpstrkey = 'max_views' (or 'max_views_add')
and the setType is always the same

fmido88 replied:
done

danielneis commented on classes/output/index.php:
```diff
@@ -60,11 +64,19 @@ public function export_for_template(renderer_base $output) {
             'hasoverrides' => !empty($overrides),
-            'newoverrideurl' => (new moodle_url('/availability/condition/maxviews/override.php', ['courseid' => $this->courseid]))->out(false),
+            'newoverrideurl' => (new moodle_url('/availability/condition/maxviews/override.php',
+                                                    ['courseid' => $this->courseid]))->out(false),
```
please, indent correctly

fmido88 replied:
predefine the url

danielneis commented on db/upgrade.php:
```diff
@@ -57,5 +54,36 @@ function xmldb_availability_maxviews_upgrade($oldversion) {
+
+        // Define the new field for timemodified.
+        $field = new xmldb_field('timeupdated', XMLDB_TYPE_INTEGER, 10, null, XMLDB_NOTNULL, null, 0, 'timecreated');
+        // Conditionally lunch add field.
```
"launch" :)

fmido88 replied:
I think I was hungry :)

danielneis commented on db/upgrade.php:
```diff
@@ -57,5 +54,36 @@ function xmldb_availability_maxviews_upgrade($oldversion) {
+
+        // Define the new field for timecreated.
+        $field = new xmldb_field('timecreated', XMLDB_TYPE_INTEGER, 10, null, XMLDB_NOTNULL, null, 0, 'maxviews');
+        // Conditionally lunch add field.
```
"launch"

danielneis commented on db/upgrade.php:
```diff
@@ -57,5 +54,36 @@ function xmldb_availability_maxviews_upgrade($oldversion) {
+        $field = new xmldb_field('overriderid', XMLDB_TYPE_INTEGER, 10, null, XMLDB_NOTNULL, null, 0, 'userid');
+        $field->setComment('The id of the user who did the overriding process.');
+
+        // Conditionally lunch add field.
```
"launch".
Haven't you generated this code with the XMLDB Editor web interface?

fmido88 replied:
No I generated it manually

danielneis commented on delete.php:
```diff
@@ -23,23 +23,43 @@
  */
 
 require_once('../../../config.php');
+require_once(__DIR__.'/classes/event/maxviews_override_deleted.php');
```
as the namespace is correctly set on the class, it shouldn't be needed.
have you tried it without this line? (after cleaning the caches)

fmido88 replied:
When I tried it for the first time, there was an error that the file not found, but you are right maybe I forgot to clean the cash.

danielneis commented on delete.php:
```diff
@@ -23,23 +23,43 @@
 require_login();
-
+$context = context_course::instance($courseid);
 $ctx = context_system::instance();
```
you can delete $ctx, please

fmido88 replied:
done

danielneis commented on index.php:
```diff
@@ -27,13 +27,13 @@
 require_login($courseid);
-
+$context = context_course::instance($courseid);
 $ctx = context_system::instance();
```
you can delete $ctx, please

fmido88 replied:
We are using it in page context, but it's ok I'll change the page context.

danielneis commented on lang/en/availability_maxviews.php:
```diff
@@ -22,24 +22,47 @@
 $string['overrides'] = 'Overrides';
 $string['overrideupdated'] = 'Override updated';
 $string['participant'] = 'Participant';
+$string['participant_help'] = 'Select the participants needed to override their maximum views for the selected course module.';
```
please change to "the participants you want to override the maximum views"

fmido88 replied:
done

danielneis commented on override.php:
```diff
@@ -23,14 +23,17 @@
 
 require_once('../../../config.php');
+require_once(__DIR__.'/classes/event/maxviews_override_updated.php');
+require_once(__DIR__.'/classes/event/maxviews_override_created.php');
```
these should't be needed as well as the namespace is correctly defined on both classes

danielneis commented on override.php:
```diff
@@ -23,14 +23,17 @@
 
 require_login($courseid);
 
 $ctx = context_system::instance();
```
plese, delete $ctx

fmido88 replied:
done

danielneis commented on override.php:
```diff
@@ -46,37 +49,104 @@
+                                ]);
+                $recordobject = (object)$newrecord;
+                // Adding the old maxviews.
+                $recordobject->maxviews += $oldrecord->maxviews;
```
I think we should be testing for override type here, right?

fmido88 replied:
I didn't create a behat test before. I'll try to do.
but I tested it manually and it worked find

danielneis commented on override.php:
```diff
@@ -46,37 +49,104 @@
+            if ($oldrecord = $DB->get_record('availability_maxviews', $record)) {
+                $newrecord = array_merge($newrecord,
+                                ['id' => $oldrecord->id,
+                                'timeupdated' => time(),
```
please, don't use array_merge, here.
you can simply assign $newrecord->id = $oldrecord->id , and $newrecord->timeupdated = time()

danielneis commented on override.php:
```diff
@@ -46,37 +49,104 @@
+                                ['id' => $oldrecord->id,
+                                'timeupdated' => time(),
+                                ]);
+                $recordobject = (object)$newrecord;
```
please, don't creat a new variable here, and cast only if really needed

danielneis commented on override.php:
```diff
@@ -46,37 +49,104 @@
+                $recordobject->maxviews += $oldrecord->maxviews;
+                // Check if there is old reset first.
+                $recordobject->lastreset = max($recordobject->lastreset, $oldrecord->lastreset);
+                $DB->update_record('availability_maxviews', $recordobject);
```
you can cast $newobject here when calling update_record

danielneis commented on override.php:
```diff
@@ -46,37 +49,104 @@
         $msg = get_string('overrideadded', 'availability_maxviews');
     }
+    // Check the type of the event.
+    if ($eventarray['other']['type'] == 'updated') {
```
you don't need an extra if, you can create the event on line after you define $eventarray[other][type]

fmido88 replied:
done

danielneis commented on templates/index.mustache:
```diff
@@ -43,6 +51,8 @@
               <td>{{userfullname}}</td>
               <td>{{maxviews}}</td>
               <td>{{#lastreset}}{{#userdate}} {{lastreset}} , {{#str}} strftimedatetime, core_langconfig {{/str}} {{/userdate}}{{/lastreset}}</td>
+              <td>{{modifier}}</td>
```
modifier or overrider?

danielneis commented on classes/output/index.php:
```diff
@@ -60,11 +64,19 @@ public function export_for_template(renderer_base $output) {
             $overrides[$key]->userfullname = fullname($o);
+
+            // The user that did the override process.
+            $modifer = \core_user::get_user($o->overriderid);
```
please, use overrider

danielneis commented on classes/output/index.php:
```diff
@@ -60,11 +64,19 @@ public function export_for_template(renderer_base $output) {
+            // The user that did the override process.
+            $modifer = \core_user::get_user($o->overriderid);
+            $overrides[$key]->modifier = fullname($modifer);
+            $overrides[$key]->date = max($o->timecreated, $o->timeupdated);
```
we can always use timeupdated, just make sure to set it on create time too.

fmido88 replied:
on the first override creation there is no timeupdated, it is equal to 0.
I'll modify it in the creation code, so timeupdated will set to time() at creation too

danielneis replied:
did you remember to always update timeupdated ? we have to remove this max call here

fmido88 replied:
Yes, in override.php in any update action (but not in creation).

danielneis replied:
please, set his field in the creation process so we do not need this max here

fmido88 replied:
Done

danielneis reviewed (changes requested):
Hello, Mo Farouk
Thank you very much for your contributions.
The good looks very good and consistent.
It only needs a few fixes and improvements to be ready to merge.
Please, take a look at the comments and if you have any questions, feel free to ask.
Best regards,
Daniel

fmido88 committed:
Add files via upload

fmido88 commented:
@danielneis All done

danielneis commented on classes/form/override.php:
```diff
@@ -40,33 +44,86 @@ public function definition() {
+            if (has_capability('moodle/course:ignoreavailabilityrestrictions', $context, $user)) {
+                continue;
+            }
+            $userfields = \core_user\fields::get_identity_fields($context, true);
```
this could be outside of the foreach loop

fmido88 replied:
You are right, I'll move it now.

danielneis reviewed (changes requested):
thanks! most of thigs were accomplished. there's just a couple of little fixes to do.
I'll try the code locally next week. If you can write some behat tests would be awesome, but there's no problem if you don't.
Best,
Daniel

fmido88 committed:
some fixes

- fix timemodified on updating old override records.
- move userfields data outside the loop in override form

fmido88 committed:
Remove max() and rely on timeupdated only

danielneis commented on override.php:
```diff
@@ -39,44 +38,109 @@
+            $newrecord = [
+                'maxviews' => $data->maxviews,
+                'overriderid' => $USER->id,
+                'timeupdated' => time(),
```
we don't have timecreated here too?

fmido88 replied:
in line 130, because I use it in inserting records only.

danielneis reviewed (changes requested):
I think we're missing timecreated on the creation of the override. could you please double check and add if needed?

danielneis commented:
I've tried the code locally and experienced a few errors.
classes/output/index.php : line 71 should be like this:
            // The user that did the override process.
            if (empty($o->overriderid)) {
                $overrides[$key]->overrider = ''; 
                $overrides[$key]->date = get_string('never');
            } else {
                $overrider = \core_user::get_user($o->overriderid);
                $overrides[$key]->overrider = fullname($overrider);
                $overrides[$key]->date = $o->timeupdated;
            }   

Because for existing overrides there are no overriderid.

I also tried to edit an override that existed for the admin user (that is enrolled in course as student) and it didn't appear on the list. we should fetch the full list of participants with no extra filter.

Then I tried to edit another existing override but got an error :
Exception - Call to a member function trigger() on null
Output buffer:  <br /> <b>Warning</b>: Undefined variable $event in <b>moodle/availability/condition/maxviews/override.php</b> on line <b>143</b><br /> 

Please, fix this errors and re-submit.
If you can add any behat tests to make sure everything is keep working would be awesome.

fmido88 committed:
Some fixes.

- Fix empty overriderid issue.
- Adding option for filtering users in override form or not.
- Define the missing update event.

fmido88 commented:
@danielneis sorry for the errors.
The first one I didn't put it in count and didn't appear to me as I worked with overriderid from the beginning so I missed that some used the plugin without this field. I fixed it.

The second problem about the filtration: I'm not convinced  about overriding a user who already can ignore restriction. Though I added an option for the filtration in the settings of the plugin. I hope this is ok.

The third problem: missing definition of $event. Fixed.

About the behat test, I'm really don't have a background about how to create a behat test.

danielneis commented:
Hello,
About the multiple user select:
At first, I've found it usefull to create multiple overrides at once.
But then, I've found it brings more problems than it solves.
When editing an override, I should not be able to select the user I'm editing and, for sure, not selecting more than one.
When I select more than one, it looks like only the first is overriden.
When I create an override for multiple users, if there are already overrides for some, they are not updated, only the non existent one is created. It should update the existent or have a way to not allow to select users that already has overrides.
In this sense, you can think about group overrides, like assignment and quiz have, would be an option. 
If you can split your code and submit only the part of the events/logs, I can integrate that.
If you can split your code and submit only the part with the setting and the ability to sum the override, I can integrate that.
In summary, if you remove this part about multiple users I can integrate the rest.
Than you can elaborate more on this topic and resubmit.
Thank you for your collaboration and patience.
Best,
Daniel

danielneis commented:
I've forgot to mention that on my last test, the "Overrided by" column was not displaying anything. The table has more headers than columns on the rest of the lines.

fmido88 commented:
Hello @danielneis 
I'm sorry for late response
I made some changes and I'll upload it soon regarding updating existence overrides.
For the "Overrided by" column I discovered a mistake in the mustache file and fixed it.

fmido88 committed:
fixes

- Fix Column override by.
- Fix update overrides on multiple selected users

fmido88 commented:
In order to show in the list of users the only ones that hasn't been overridden yet, I think we must make a multiple pages form, The first one is to select the course modules and the second one for users and the rest of the form.

We can also add settings for either bulk override or just single user selection.

fmido88 committed:
Add files via upload

fmido88 committed:
Add files via upload

- Remove multiple selection of uses in case of editing.

fmido88 committed:
Add files via upload

fmido88 committed:
Merge branch 'main' into main

fmido88 committed:
callback function

- Adding a callback function to display the views limits in cms

fmido88 committed:
Check the kind

Check the kind (cm or section) before get the overrides

fmido88 committed:
lang strings for display remained views

fmido88 commented:
Hello @danielneis I hope you are doing well.
Please check the latest updates.

fmido88 committed:
Add files via upload

fmido88 committed:
Better logic for display

- Better logic for displaying view limits.
- Make it optional in setting page.

fmido88 committed:
Fix course module filtration

fmido88 committed:
create amd module

fmido88 committed:
display views limits on course view page only

fmido88 committed:
add description and copyright

fmido88 committed:
Fix classes names

-Fix classes names for info_module and info_section
- create seperated function to get the views count and views limit

fmido88 committed:
use info classes

fmido88 committed:
separated function to extract maxviews

fmido88 committed:
use observer instate of time interval

fmido88 committed:
Limit the observer on adding nodes only

fmido88 committed:
Fix error due to overrides of deleted cms

fmido88 committed:
latest

fmido88 committed:
fix lang

fmido88 committed:
"Compatibility for 4.4"

fmido88 committed:
Update the logic concerning display maxviews before restriction
